airflow/providers/edge/executors/edge_executor.py

```diff
@@ -151,6 +151,18 @@ def cleanup_stuck_queued_tasks(self, tis: list[TaskInstance]) -> list[str]:  # p
         """
         raise NotImplementedError()
 
+    def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[TaskInstance]:
+        """
+        Try to adopt running task instances that have been abandoned by a SchedulerJob dying.
+
+        Anything that is not adopted will be cleared by the scheduler (and then become eligible for
+        re-scheduling)
+
+        :return: any TaskInstances that were unable to be adopted
+        """
+        # We handle all running tasks from the DB in sync, no adoption logic needed.
```

Comment:
Wait, I'm only now realizing this. So every executor (in a multiple scheduler/HA environment) is reading all tasks across the airflow cluster? This is another unusual sideeffect of how you've implemented this executor. Usually executors maintain a queue of tasks that they've received from _only_ their scheduler, but in this case your executor implementation is going to be stepping over the toes of every other instance of the edge executor in an HA environment.

Is this intended?

Reply:
For the MVP it is intended.

I would still refer to the fact that this is MVP. Main intent is to have something running and then to optimize.
I would not say the schedulers are stepping on each-others toe, they just share one DB table for all open jobs. Jobs are purged from the table a few minutes after completion. So it is all very simple short running transactions and all should be in cache of Postgres. The shared table contains N*PARALLELSIM open records for N schedulers.

For AIP for MVP I documented that max 100 workers are the planned start. Later improvement:
![image](https://github.com/user-attachments/assets/7f6bf38d-940a-4bf0-bdd2-caec5d6e66ad)

But as there might be doubt... just created a small DAG with a dynamic mapped PythonOperator, 1 print() statement. Mapped to 500 tasks. Executed on my laptop with webserver, 1 scheduler, 4 worker (each concurrency of 8) and the 500 tasks were completed in 11:48min. Whereas almost all time was used by python processes forked off workers to start/init. Second biggest CPU consumer was webserver (~1 core) for the API backend. Scheduler as well as Postgres DB each consumed ~3-5% of one of the cores.
Have not more hardware at my hands to distribute more load by workers but scheduler in this small test was not showing any performance challenge.

Reply:
@o-nikolas Had a longer thought on this and as I am still seeking for other reviewers... two arguments towards "every executor is reading all tasks across the cluster":

1. During scheduling (which is the rather performance critical activity) just DB entries are added. SQLalchemy will batch them up and flush them in the scheduling loop. This should be pretty efficient.
2. All tasks are _only_ read from the table in the `sync()` call in order to clean up the internal state and the table itself. `sync()` is called periodically and as it is a cleanup it is mainly to remove old tasks and check for completion (to schedule more). Besides binding tasks to a specific scheduler instance we could also reduce the frequency of cleanup to say for example only clean-up every N `sync()` calls. I saw this being executed ~every second. Therefore I took a look to (current) timing on my laptop and came to:
  - If 1 task is in the edge_job table then the `sync()` needs 1-2ms per call (which seems to happen roughly every second)
  - If I added 1000 artificial tasks the time for `sync()` increased to 25-30ms.
  - Therefore I took a look to the code, optimized it one round in https://github.com/apache/airflow/pull/42048/commits/c2672ed2ac356de6f925f1be9474d4fba678ae48 which filters the rows for relevance in a first optimization to the relevant states for the check in `sync()`. With this if 1000 tasks are queued the time drops down to 2-3ms and will only increase with the amount of running or tasks to clean.
3. If you compare the "additional DB load" to Celery - also in Celery per task tuples are added and updated in the result table. Did not count the number of transactions but it is also one row per task in a global table `celery_taskmeta`. With Edge Executor we are in a similar range.

Does this improve your opinion about DB overhead and performance? Otherwise please let me know if you would request me to filter rows per scheduler down already in MVP. (We currently use 3 scheduler instances, in my view the overhead is quite minimal for the sync).

Reply:
Two thoughts:

1) You've taken great care in this case to make the DB interaction fairly optimized. Thanks for that! But a big concern is that we're setting a _precedent_ here, for folks to use the DB as the queue for tasks in their executors. With multiple executors running concurrently and others perhaps not taking as much care as you have, could lead to poorer outcomes. I think as a community and piece of software, we are agreed on removing as many of our direct dependencies on the metadata DB and strongly encouraging contributors to not write code that expects to have direct access to it like this. I think this executor as implemented works against that goal.

2) Not only the DB impact, but the logical impact: up until this point each executor/scheduler pair owned a specific set of tasks. And there are assumptions in Airflow logic based on that. This implementation means that all executors are sharing tasks because they are communicating over the DB. I can't sign off on the implications of this honestly, only that this is not the norm for executors and I'm not sure what the downstream effects of that could be.

Happy to have this PR merged, I'd just be curious to see if others have the same concerns. If not, then feel free to get approvals and merge! :smiley:

Reply:
As of DB impact concerns I made a load test yesterday in our Azure AKS environment and used K8s PODs for workers inside the cluster. Based on the current code my results were:

Setup:
- 3 Scheduler (PODs)
- 30 Edge Worker, each concurrency=6
- Postgres DB, 8 Cores, 32GB RAM (Azure Flexible Server)
- 1 DAG with 1 Mapped Task, 1000 mappings. Second DAG which calls the DAG i a batch

Results:
- Constant execution needed 2:30min per 1000 tasks/DAG
- DB Load was up ~200 IOPS and +20% CPU over the base load
- 3 Scheduler PODs each needed ~35% of a single core
- Load was very constant over multiple hours, would have achieved 550 000 tasks in a day

Main limiting factor and driver for DB load seems to be rather the Edge Worker making a lot of API requests --> DB via JSON RPC. Needed to scale Webserver and PGBouncer following as of the DB queries mainly from Webserver.

I would not have any fears in regards of DB load seen from monitoring that was emitted by Scheduler itself. Rather the "chattiness" from the AIP-44 internal API will need to be replaced with upcoming AIP-72 if there should be any scaling promise. But of course the Scheduler also can be optimized... post MVP.

Reply:
@o-nikolas I added explicit notes into the RST of the executor that a performance assessment ensuring no side effect/impact on other executors before v1.0.0 release.. OK then to merge?

Reply:
Sure, no worries! Regarding my points 1 and 2 above, it's not entirely about performance. But I'm glad you got some other points of view on this one (assuming others reviewed this and silently agreed it's okay).

Reply:
Also, I'm working on some multi-team stuff lately and realized that the way this executor shares tasks between different instances will likely not work for that kind of setup. In multi-team executors should be siloed to a specific team, and should not have access to the tasks from other teams. This executor will definitely require updates to support being used in a multi-team capacity.

Reply:
Thanks for the heads-up. Yes I am sure that we can not this as a limitation for multi-team. As it is a MVP, not all use cases are supported. If it matures and this will be a requirement that is needed I am fine to extend this. Also for large scaling maybe alternative queues to DB can be integrated. Not limited to stay as it is and I am committed to improve!